docs/src/topics/assembly.md

```diff
@@ -83,3 +83,310 @@ for t in 1:timesteps
+    `K = spzeros(ndofs(dh), ndofs(dh))` and then inserting entries is excruciatingly slow
+    due to the sparse data structure so this method is not even considered.
+
+For the comparison we need a representative global matrix to assemble into. In the following
+setup code we create grid with traingles and a DofHandler with two fields (a quadratic
+vector field and a linear scalar field). From this we instantiate the global matrix.
+
+```@example assembly-perf
+using Ferrite
+
+# Interpolations
+ip_u = Lagrange{RefTriangle, 2}()^2 # Quadratic vector field
+ip_p = Lagrange{RefTriangle, 1}()   # Linear scalar field
+
+# DofHandler with two fields
+const N = 100
+grid = generate_grid(Triangle, (N, N))
+const dh = DofHandler(grid)
+add!(dh, :u, ip_u)
+add!(dh, :p, ip_p)
+close!(dh)
+
+# Global matrix and a corresponding assembler
+const K = allocate_matrix(dh)
+nothing # hide
+```
+
+#### Strategy 1: matrix indexing
+
+The first strategy is to index directly, using the vector of global dofs, into the global
+matrix:
+
+```@example assembly-perf
+function assemble_v1(_, K, dofs, Ke)
+    K[dofs, dofs] += Ke
+    return
+end
+nothing # hide
+```
+
+This looks very simple but it is very inefficient (as the numbers will show later). To
+understand why the operation `K[dofs, dofs] += Ke` (with `K` being a sparse matrix) is so
+slow we can dig into the details.
+
+In Julia there is no "`+=`"-operation: `x += y` is identical to `x = x + y`. Translating
+this to our example we have
+
+```julia
+K[dofs, dofs] = K[dofs, dofs] + Ke
+```
+
+We can break down this a bit further by introducing some temporary variables:
+
+```julia
+tmp1 = K[dofs, dofs]   # 1
+tmp2 = tmp1 + Ke       # 2
+K[dofs, dofs] = tmp2   # 3
+```
+
+Now the problem with this strategy becomes a bit more obvious.
+ - In line 1 there is first an allocation of a new matrix (`tmp1`) followed by indexing into
+   `K` to copy elements from `K` to `tmp1`. Both of these operations are rather costly:
+   allocations should always be minimized in tight loops, and indexing into a sparse matrix
+   is non-trivial due to the data structure. In addition, since `dofs` vector contains the
+   global indices (which are not sorted nor consecutive) we have random access patterns.
+ - In line 2 there is another allocation of a matrix (`tmp2`) for the result of the addition
+   of `tmp1` and `Ke`.
+ - In line 3 we again need to index into the sparse matrix to copy over the elements from
+   `tmp2` to `K`. This essentially duplicates the indexing effort from line 1 since we
+   lookup the same locations in `K` again.
+
+#### Strategy 2: scalar indexing
+
+A variant of the first strategy is to explicitly loop over the indices and add add the
+elements individually as scalars:
+
+```@example assembly-perf
+function assemble_v2(_, K, dofs, Ke)
+    for (i, I) in pairs(dofs)
+        for (j, J) in pairs(dofs)
+            K[I, J] += Ke[i, j]
+        end
+    end
```

Comment:
Based on the timings this is a perfectly good way to assemble so it could be used in the very first tutorial if one doesn't want to introduce the assembler already there (as suggested by Chris during FerriteCon).

Reply:
Maybe we should also link to this topic section in the first tutorial?